From 9fa57589a61b49fef5963f8b757212a767c31d07 Mon Sep 17 00:00:00 2001
From: Stelios Tsampas <loathingkernel@gmail.com>
Date: Sun, 28 Jun 2026 22:52:59 +0300
Subject: [PATCH 4/8] upscalers: update handling for FSR4 4.1.1

---
 upscalers.py | 30 ++++++++++++++++--------------
 1 file changed, 16 insertions(+), 14 deletions(-)

diff --git a/upscalers.py b/upscalers.py
index 754f703..a455146 100644
--- a/upscalers.py
+++ b/upscalers.py
@@ -4,7 +4,6 @@ import hashlib
 import json
 import lzma
 import os
-import shutil
 import urllib.request
 from functools import lru_cache
 from urllib.error import HTTPError, URLError
@@ -80,8 +79,8 @@ def __dll_download_exists(url: str) -> bool:
 
 __dlss_section = 'dlss_files'
 __xess_section = 'xess_files'
-__fsr3_section = 'fsr3_files'
 __fsr4_section = 'fsr4_files'
+__ffx3_section = 'ffx3_files'
 __version_file = 'upscaler_files'
 
 
@@ -110,7 +109,7 @@ def __get_xess_dlls(version: str = 'default') -> dict:
     }
 
 
-def __get_fsr3_dlls(version: str = 'default') -> dict:
+def __get_ffx3_dlls(version: str = 'default') -> dict:
     return {
         'drive_c/windows/system32/umu/amd_fidelityfx_vk.dll': __get_dll_manifest(
             'fsr_31_vk', version
@@ -133,8 +132,8 @@ def __get_upscaler_items(name: str, version: str) -> tuple[dict, Callable, str]:
     upscalers = {
         'dlss': (__get_dlss_dlls, __download_extract_zip, __dlss_section),
         'xess': (__get_xess_dlls, __download_extract_zip, __xess_section),
-        'fsr3': (__get_fsr3_dlls, __download_extract_zip, __fsr3_section),
         'fsr4': (__get_fsr4_dlls, __download_extract_zip, __fsr4_section),
+        'ffx3': (__get_ffx3_dlls, __download_extract_zip, __ffx3_section),
     }
     get_items, dlfunc, section = upscalers[name]
     try:
@@ -419,17 +418,20 @@ def setup_upscalers(
     """
     dlss_version = get_version(env, 'PROTON_DLSS_UPGRADE', 'default')
     xess_version = get_version(env, 'PROTON_XESS_UPGRADE', 'default')
-    fsr3_version = get_version(env, 'PROTON_FSR3_UPGRADE', '1.0.1.41314')
-    fsr4rdna3_version = get_version(env, 'PROTON_FSR4_RDNA3_UPGRADE', '4.0.0')
+
+    fsr3_version = get_version(env, 'PROTON_FSR3_UPGRADE', 'default')
     fsr4_version = get_version(env, 'PROTON_FSR4_UPGRADE', 'default')
-    fsr4_version = fsr4rdna3_version if 'fsr4rdna3' in compat_config else fsr4_version
+    fsr4_version = fsr3_version if 'fsr3' in compat_config else fsr4_version
+
+    ffx3_version = get_version(env, 'PROTON_FFX3_UPGRADE', '1.0.1.41314')
 
     upscaler_replace = set()
     upscalers = (
         ('dlss', dlss_version, 'dlss' in compat_config),
         ('xess', xess_version, 'xess' in compat_config),
-        ('fsr3', fsr3_version, 'fsr3' in compat_config),
-        ('fsr4', fsr4_version, 'fsr4' in compat_config or 'fsr4rdna3' in compat_config),
+        # amdxcffx64 4.1.1
+        ('fsr4', fsr4_version, True),
+        ('ffx3', ffx3_version, 'ffx3' in compat_config),
     )
     for upscaler in upscalers:
         name, version, enabled = upscaler
@@ -438,11 +440,12 @@ def setup_upscalers(
             upscaler_replace.add(name)
 
     if 'fsr4' in upscaler_replace:
-        env['FSR4_UPGRADE'] = '1'
+        if 'fsr3' in compat_config:
+            env['FSR3_UPGRADE'] = '1'
+        elif 'fsr4' in compat_config:
+            env['FSR4_UPGRADE'] = '1'
         if 'mlfg' in compat_config:
             env['MLFG_UPGRADE'] = '1'
-        if 'fsr4rdna3' in compat_config:
-            env['DXIL_SPIRV_CONFIG'] = 'wmma_rdna3_workaround'
 
     if 'dlss' in upscaler_replace:
         env.setdefault(
@@ -457,7 +460,7 @@ def setup_upscalers(
     if 'xess' in upscaler_replace:
         pass
 
-    if 'fsr3' in upscaler_replace:
+    if 'ffx3' in upscaler_replace:
         pass
 
     if upscaler_replace:
@@ -465,5 +468,4 @@ def setup_upscalers(
         log.debug(f'WINE_UPSCALER_REPLACE: {env["WINE_UPSCALER_REPLACE"]}.')
 
 
-
 __all__ = ['setup_upscalers']
-- 
2.55.0

